File: Generating/prototype/javascript/rectangle.js
const Shape = require('./shape')

class Rectangle extends Shape {
  constructor(params) {
    super(params)
    if (params) {
      this.width = params.width
      this.height = params.height
    }
  }

  clone() {
    return new Rectangle(this)
  }

  square() {
    return this.width * this.height
  }

  equals(object) {
    switch (true) {
      case (object.x !== this.x):
        return false
      case (object.y !== this.y):
        return false
      case (object.color !== this.color):
        return false
      case (object.height !== this.height):
        return false
      case (object.width !== this.width):
        return false
      default:
        return true
    }
  }
}

module.exports = Rectangle

File: Generating/prototype/cpp/rectangle.cpp
#ifndef rectangle_impl
#define rectangle_impl

#include "rectangle.hpp"
#include "shape.cpp"
#include <string>

Rectangle::Rectangle(int x, int y, int width, int height, string color)
{
  setX(x);
  setY(y);
  setColor(color);
  setWidth(width);
  setHeight(height);
};

Rectangle * Rectangle::clone() const
{
  return new Rectangle(*this);
};

int Rectangle::square()
{
  return getWidth() * getHeight(); 
};

bool Rectangle::equals(Rectangle targetObject)
{
  if (targetObject.getX() != x) {
    return false;
  } else if (targetObject.getY() != y) {
    return false;
  } else if (targetObject.getColor() != color) {
    return false;
  } else if (targetObject.getWidth() != width) {
    return false;
  } else if (targetObject.getHeight() != height) {
    return false;
  } else {
    return true;
  }
};

void Rectangle::setParams(int new_x, int new_y, int new_width, int new_height, string new_color)
{
  setX(new_x);
  setY(new_y);
  setColor(new_color);
  setWidth(new_width);
  setHeight(new_height);
};

int Rectangle::getWidth()
{
  return width;
};

int Rectangle::getHeight()
{
  return height;
};

void Rectangle::setWidth(int new_width)
{
  width = new_width;
};

void Rectangle::setHeight(int new_height)
{
  height = new_height;
};

#endif